src/qutip_qip/decompose/decompose_two_qubit_gate.py

```diff
@@ -0,0 +1,293 @@
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _decompose_to_two_level_arrays(input_gate):
+    """Output of two qubit gate in terms of two-level numpy arrays.
+    """
+    check_gate(input_gate, num_qubits=2)
+    input_array = input_gate.full()
+
+    # Calculate the two level numpy arrays
+    array_list = []
+    index_list = [[0, 1], [0, 2], [0, 3], [1, 2], [1, 3]]
+    for i in range(len(index_list)):
+        index_1, index_2 = index_list[i]
+
+        # Values of single qubit U forming the two level unitary
+        a = input_array[index_1][index_1]
+        a_star = np.conj(a)
+        b = input_array[index_2][index_1]
+        b_star = np.conj(b)
+        norm_constant = cmath.sqrt(np.absolute(a*a_star)+np.absolute(b*b_star))
+
+        # Create identity array and then replace with above values for index_1
+        # and index_2
+        U_two_level = np.identity(4, dtype=complex)
+        U_two_level[index_1][index_1] = a_star/norm_constant
+        U_two_level[index_2][index_1] = b/norm_constant
+        U_two_level[index_1][index_2] = b_star/norm_constant
+        U_two_level[index_2][index_2] = -a/norm_constant
+
+        # Change input by multiplying by above two-level
+        input_array = np.dot(U_two_level, input_array)
+
+        # U dagger to calculate the gates
+        U__two_level_dagger = np.transpose(np.conjugate(U_two_level))
+        U__two_level_dagger = Qobj(U__two_level_dagger, dims=[[2, 2], [2, 2]])
+        array_list.append(U__two_level_dagger)
+
+    # for U6 - multiply input array by U5 and take dagger
+    U6_dagger = input_array
+    array_list.append(Qobj(U6_dagger, dims=[[2, 2], [2, 2]]))
+    return(array_list)
+
+
+def decompose_two_qubit_to_two_level_unitary(input_gate):
+    """Decomposes input two-qubit unitary into two-level unitary gate objects.
+    Returns
+    ----------
+    tuple: (two_level_dictionary, two_level_gates)
+    two_level_dictionary
+        Dictionary containing string labels to insert user-defined gates in
+        a quantum circuit.
+    two_level_gates
+        Gates describing the decomposition of the two-qubit input gate.
+    """
+    check_gate(input_gate, num_qubits=2)
+    array_list = _decompose_to_two_level_arrays(input_gate)
+    user_gates = {
+        "U1": array_list[0],
+        "U2": array_list[1],
+        "U3": array_list[2],
+        "U4": array_list[3],
+        "U5": array_list[4],
+        "U6": array_list[5],
+        }
+
+    U1_gate = Gate("U1", targets=[0, 1])
+    U2_gate = Gate("U2", targets=[0, 1])
+    U3_gate = Gate("U3", targets=[1, 0])
+    U4_gate = Gate("U4", targets=[0, 1])
+    U5_gate = Gate("U5", targets=[0, 1])
+    U6_gate = Gate("U6", targets=[0, 1])
+
+    gate_list = (U6_gate, U5_gate, U4_gate, U3_gate, U2_gate, U1_gate)
+
+    return(user_gates, gate_list)
+
+
+def decompose_two_qubit_to_CNOT_and_single_qubit_gates(input_gate):
```

Comment:
Output of this function is not what it is expected to be. For this reason, all of the smaller gates for U6,...,U1 have been defined separately. A for loop will be created for this once the output matches exactly to the input. In addition, all the two qubit decomposition methods will be made private after a general two qubit decomposition function isalso be added - provides choice for circuit output. 

Need to recalculate the target and control qubits because they were different from what was expected for the [two level circuit output function](https://github.com/qutip/qutip-qip/pull/70/files#diff-4fd62396e554b500c5f57c03c2e6cf0123b85fb4ecb549a57b97f8d120a25f62R174). The output of the two level output matches [exactly to the input](https://github.com/qutip/qutip-qip/pull/70/files#diff-5b0436268788ff97a67b84a89049cd3120d5e7f7013de9df7073097ed2046adfR13) with new values for controls and targets. 

Just because I know the correct control and targets does not imply how the controls and targets for CNOT and Pauli X of each gate should be corrected. 

The hope is to check and compare my calculations with the targets and controls from gray code function's output. 

Tentative new values for the target/controls - [cnot_gates_2_qubits.pdf](https://github.com/qutip/qutip-qip/files/6859753/cnot_gates_2_qubits.pdf)

Reply:
> Issue with gray code - There's ambiguity as to how you map from 1 state to another as long as only 1 bit is changed at each step.

I think if all qubits are the same, the physical resources shouldn't make a difference. Of course in real hardware qubits are not the same and gate implementation is not always symmetric so one way might be better than the other. But I think that is a very concrete scenario and we don't need to care about it now. It would be good enough to just define a convention like *always go with the route for smaller binary number (01 < 10)*.

>The binary sequence for 2 qubits is binary_sequence= [00, 01, 10, 11] which can be re-written in terms of array indices as [0, 1, 2, 3].

Not sure if this is helpful here, `numpy` has `ravel_multi_index` and `unravel_index` to compute between binary representation and matrix row/column index.

Reply:
E.g. 
```python
>>> np.ravel_multi_index([0,1,0],[2,2,2])
2
```

Reply:
> the physical resources shouldn't make a difference

But the order that you map in does matter because the sequence has to be cyclic. When I did my calculations, I had tried to use the shortest method to go from one state another in the hope that the total gate number would decrease. As long as you choose a consistent method ( for this project, I have chosen [reflect and prefix method](http://www.cs.williams.edu/~jannen/teaching/f16/cs135/lectures/lecture.11.pdf)). 

I now think my error is not related to the gray code sequence but instead about the order of gates and/or controls, targets (see example below). I might have missed some information from the reference papers. Now, my plan is to at least have a working code for what was planned this week and then come back to correcting a few lines after I have figured out my error. This should also give you a chance to review the code logic etc.  

Suppose I want to decompose the following matrix :

![image](https://user-images.githubusercontent.com/66048318/126737885-19c46b84-bcc8-4ea1-83e2-a842cae8a8ba.png)

The non-trivial basis vectors are 01 and 10. Using the gray code generator, 01 -> 11 -> 10. 

- For 01 -> 11 : When qubit 2 is 1, qubit 1 is flipped. Thus, there's a CNOT gate added to the circuit with `controls=[1], targets=[0], control_value=1`
- For 11 -> 10 : When qubit 1 is 1, qubit 2 is flipped. This is a controlled gate on target 0 with control value 1. But because this is the last steps in the sequence, the non-trivial sub-matrix has to be added to the circuit. The form of this is U6 whose decomposition is known in Lemma 5.1.  
- The last gate is to go back to state 01. 

![image](https://user-images.githubusercontent.com/66048318/126738533-75473c64-1fbf-40f6-a68e-251fce611293.png)

![image](https://user-images.githubusercontent.com/66048318/126739040-13c65c32-200e-40fd-84cb-554ffe18c4dc.png)


The circuit's unitary is calculated as :

![image](https://user-images.githubusercontent.com/66048318/126738488-ca67a860-064c-4076-ab98-3a886a52b16c.png)

Reply:
> E.g.
> 
> ```python
> >>> np.ravel_multi_index([0,1,0],[2,2,2])
> 2
> ```

Thanks ! I defined those functions yesterday. I'll keep them for now. If it's preferred to use numpy, I'll make the changes then.

Reply:
Looks good.

![image](https://user-images.githubusercontent.com/12125783/126755279-3e1f6072-067a-4e7e-8d95-994e7b736221.png)

This circuit makes sense. This is the same way how the SWAP gate was decomposed. Only with a unitary in the middle.